### PR TITLE
Apply testfixtures

### PR DESCRIPTION
**Pull request by damhuonglan**
### What does this PR do?
Standardize unittest by replace current log capture and output capture with testfixtures', assertStrEqual with built-in assertMultiLineEqual.

### Description
ph5/core/tests/test_base.py provide:
 * Function change_logger_handler() : to prevent logs to be printed to console. This will be put in setUpClass() of the tests that have log printed out to screen.
 * Fuction revert_logger_handler(): return logger to original status. This will be put in tearDownClass() of the tests that have change_logger_handler() in setUpClass().
 * class PH5TestCase will call change_logger_handler() in setUpClass() and revert_logger_handler() in tearDownClass so any class that need to prevent logs to be printed to console can extend this class.

assertStrEqual is a function written to compare two long strings to help user identify at which position the two strings are different. However I just figured out that the built-in assertMultiLineEqual can serve that purpose too and works more stable. So I replace assertStrEqual with assertMultiLineEqual.

### CODE REVIEW

**Comment by dsentinel** [comment]
This is a first attempt at using `testfixtures` to capture logging and stdout/err messages in tests.

@damhuonglan Can you please describe the problem you're having?

**Comment by damhuonglan** [comment]
At first, using textfitures inside ph5 doesn't capture anything while inside unittest_ex it work.
But then  after I get logger from ph5 it works.

**Comment by dsentinel** [comment]
> At first, using textfitures inside ph5 doesn't capture anything while inside unittest_ex it work.
> But then after I get logger from ph5 it works.

It works, so you figured it out?

**Comment by damhuonglan** [comment]
When I worked on my laptop from home, there were some problem that I cannot get the logger from ph5. I will show it to you on Monday. Now can you compare between StringIO and testfixtures to see if I should apply that to other PRs

**Comment by damhuonglan** [comment]
There is one thing I don't like in log capture in testfixtures is I have to check the whole log. I can't just check a part of the log. I will try to see if there is a way to do that.

**Comment by damhuonglan** [comment]
@dsentinel I just push a commit that uses testfixtures to test log in more details. Let me know if there is anything I should change in the codes. And also let me know when you think I should apply this to other PR.

File: ph5/utilities/tests/test_obspytoph5.py [hunk 1]
```diff
@@ -9,10 +9,11 @@
...
 from mock import patch
+from ph5.core.tests.test_base import PH5TestCase
+from testfixtures import OutputCapture
 
 
-class TestObspytoPH5(unittest.TestCase):
-
+class TestObspytoPH5(PH5TestCase):
     def setUp(self):
         self.path = 'ph5/test_data/miniseedph5'
```
[the remark this answers is not in the document]
Comment (damhuonglan) [reply ph5/utilities/tests/test_obspytoph5.py]: I didn't overriding setUp(). The one in super class is setUpClass() which is a different one.

File: ph5/utilities/tests/test_segd2ph5.py [hunk 2]
```diff
@@ -6,28 +6,20 @@
...
         return EX
 
     def setUp(self):
-        # capture log string into log_capture_string
-        self.log_capture_string = StringBuffer()
-        logger.removeHandler(CH)
-        ch = logging.StreamHandler(self.log_capture_string)
-        logger.addHandler(ch)
-
         # create tmpdir
```
[the remark this answers is not in the document]
Comment (damhuonglan) [reply ph5/utilities/tests/test_segd2ph5.py]: I am thinking about applying tempdir stuff independently with logging handler stuff. Because sometimes we don't need to use both. May be just a function to call.

File: ph5/utilities/tests/test_segd2ph5.py [hunk 3]
```diff
@@ -6,28 +6,20 @@
...
 
-class TestSegDtoPH5(unittest.TestCase):
+class TestSegDtoPH5(PH5TestCase):
     def initialize_ph5(self, editmode):
         EX = experiment.ExperimentGroup(nickname="master.ph5")
         EX.ph5open(editmode)
         EX.initgroup()
         return EX
 
     def setUp(self):
```
[the remark this answers is not in the document]
Comment (damhuonglan) [reply ph5/utilities/tests/test_segd2ph5.py]: What do you mean super() here. Do you mean put this stuff in testbase?

File: ph5/core/tests/test_ph5api.py [hunk 4]
```diff
@@ -4,10 +4,10 @@
...
 from ph5.core import ph5api
+from ph5.core.tests.test_base import PH5TestCase
 
 
-class TestPH5API(unittest.TestCase):
-
+class TestPH5API(PH5TestCase):
     def setUp(self):
         self.ph5API_object = ph5api.PH5(path='ph5/test_data/ph5',
                                         nickname='master.ph5')
```
[the remark this answers is not in the document]
Comment (damhuonglan) [reply ph5/core/tests/test_ph5api.py]: I didn't overriding setUp(). The one in super class is setUpClass() which is a different one.

File: ph5/utilities/tests/test_metadatatoph5.py [hunk 5]
```diff
@@ -11,10 +11,11 @@
...
 from mock import patch
+from ph5.core.tests.test_base import PH5TestCase
+from testfixtures import OutputCapture
 
 
-class TestMetadatatoPH5(unittest.TestCase):
-
+class TestMetadatatoPH5(PH5TestCase):
     def setUp(self):
         self.path = 'ph5/test_data/miniseedph5'
```
[the remark this answers is not in the document]
Comment (damhuonglan) [reply ph5/utilities/tests/test_metadatatoph5.py]: I didn't overriding setUp(). The one in super class is setUpClass() which is a different one.

File: ph5/utilities/tests/test_segd2ph5.py [hunk 6]
```diff
@@ -6,24 +6,28 @@
...
-        # create tmpdir
-        self.home = os.getcwd()
-        self.tmpdir = tempfile.mkdtemp() + "/"
-        os.chdir(self.tmpdir)
+        super(TestSegDtoPH5, self).setUp()
```
Comment (dsentinel) [inline ph5/utilities/tests/test_segd2ph5.py]: `super()` call with no args.

File: ph5/utilities/tests/test_segd2ph5.py [hunk 7]
```diff
@@ -6,24 +6,28 @@
...
+        # capture log string into log_capture_string
+        self.log_capture_string = StringBuffer()
+        logger.removeHandler(CH)
+        ch = logging.StreamHandler(self.log_capture_string)
+        logger.addHandler(ch)
```
Comment (dsentinel) [inline ph5/utilities/tests/test_segd2ph5.py]: This is all done in the base class `LogTestCase`

File: ph5/core/tests/test_base.py [hunk 8]
```diff
@@ -1,7 +1,42 @@
...
+    newch = logging.StreamHandler(log)
+    logger.addHandler(newch)
+
+
+def revert_logger_handler():
+    # disable propagating to higher loggers
+    logger.propagate = 0
+    # revert logger handler
+    logger.removeHandler(newch)
+    logger.addHandler(ch)
```
Comment (dsentinel) [inline ph5/core/tests/test_base.py]: Move this into `LogTestCase` `setUpClass` and `tearDownClass`

File: ph5/utilities/tests/test_segd2ph5.py [hunk 9]
```diff
@@ -4,16 +4,17 @@
 import unittest
 import os
 import sys
+import shutil
+import tempfile
```
Comment (dsentinel) [inline ph5/utilities/tests/test_segd2ph5.py]: Why add these? Are they used?

File: ph5/utilities/tests/test_segd2ph5.py [hunk 10]
```diff
@@ -4,16 +4,17 @@
...
 import logging
 from StringIO import StringIO as StringBuffer
 from mock import patch
-from ph5 import logger, ch as CH
 from ph5.utilities import segd2ph5, tabletokef
 from ph5.core import experiment, segdreader
-from ph5.core.tests.test_base import TempDirTestCase
+from ph5.core.tests.test_base import LogTestCase, import TempDirTestCase
```
Comment (dsentinel) [inline ph5/utilities/tests/test_segd2ph5.py]:
```
from ph5.core.tests.test_base import LogTestCase, TempDirTestCase
```

File: ph5/utilities/tests/test_segd2ph5.py (hunk 9 above)
Comment (damhuonglan) [reply ph5/utilities/tests/test_segd2ph5.py]: I resolved this on the webpage. It was hard to check on the web. Then I pull to my computer to check. Check out the new commit

File: ph5/utilities/tests/test_segd2ph5.py (hunk 10 above)
Comment (damhuonglan) [reply ph5/utilities/tests/test_segd2ph5.py]: I resolved this on the webpage. It was hard to check on the web. Then I pull to my computer to check. Check out the new commit

File: ph5/core/tests/test_base.py [hunk 11]
```diff
@@ -1,8 +1,34 @@
+from ph5 import logger, ch
```
[the remark this answers is not in the document]
Comment (damhuonglan) [reply ph5/core/tests/test_base.py]: It says in your link: 
"It's okay to say this though:

from subprocess import Popen, PIPE"

[the remark this answers is not in the document]
Comment (dsentinel) [reply ph5/core/tests/test_base.py]: That's fine. 
Put stdlib first, blank line, 3rd party, blank line, internal

File: ph5/utilities/tests/test_segd2ph5.py (hunk 6 above)
Comment (damhuonglan) [reply ph5/utilities/tests/test_segd2ph5.py]: why? It works this way. If I remove the arguments from super(), it gives me the error:
TypeError: super() takes at least 1 argument (0 given)

Comment (dsentinel) [reply ph5/utilities/tests/test_segd2ph5.py]: Sorry, you are correct for python 2.7

**Comment by dsentinel** [comment]
Why did you add the `changedir` argument to `TempDirTestCase.setUp`

**Comment by damhuonglan** [comment]
Because you ask me to use tmpdir instead of self.path = 'ph5/test_data/miniseedph5' for test_obspytoph5.py and test_metadatatoph5.py. Those weren't written by me. I don't see it change dir anywhere. So my solution is apply tmpdir for those files but not change the current directory to tmpdir like others. That's why I add changedir.

**Comment by damhuonglan** [comment]
Let me try not to use that. When I am done I will ask you to review again.

**Comment by dsentinel** [comment]
damhuonglan:
Without the changedir | With changedir
-- | --
the working directory will be tmpdir, so all files created will be in this directory | the working directory will be PH5 as original
at lines 66,67, the path to miniseed folder need to be absolute path which is too long:'/Users/ldam/Documents/GIT/PH5/ph5/test_data/miniseed/' | miniseed folder is under PH5, relative path can be used:'ph5/test_data/miniseed/'
=> Instead of '/Users/ldam/Documents/GIT/PH5/ph5/test_data/miniseed/0407LOG.ms', the raw_file_name_s will turn into '/Users/ldam/Documents/GIT/PH5/ph'. | The path will be kept unchanged: "ph5/test_data/miniseed/0407LOG.ms"
=> the check at line 83, 86 failed | => the check passed

**Comment by dsentinel** [comment]
Addressing the comments in the table above:

1. So the idea here is that all the files created will be in the tempdir.

2. You can store the previous cwd as a instance attribute to be referenced for the relative paths.
Or the test_data root can be stored in `test_base.py` for all to use.
Too long for what?

3. Sorry I don't understand, this one may be clipped.

4. What file? Can the test be adapted to use this correctly?

**Comment by damhuonglan** [comment]
It is too long to save into ph5 so it will be truncated. The error I got is:
AssertionError: '/Users/ldam/Documents/git/PH5/ph5/test_data/miniseed/0407HHN.m' != '/Users/ldam/Documents/git/PH5/ph'
Another way to do it is to create tmpdir under PH5. I will implement it and push. Let see if you will agree with it.

**Comment by damhuonglan** [comment]
In ph5/core/columns.py line 273, raw_file_name_s is only 32, so you can see in the original check at line 82 has the 's' cut off at the end  too:
'ph5/test_data/miniseed/0407HHN.m' in stead of 'ph5/test_data/miniseed/0407HHN.ms'

When I try to put tmpdir under PH5/, the path still long, I will try to put under PH5/ph5/test_data.

File: ph5/utilities/tests/test_obspytoph5.py [hunk 12]
```diff
@@ -78,26 +80,21 @@ def test_main(self):
...
+        self.assertEqual('../miniseed/0407HHN.ms',
+                         ret[0]['raw_file_name_s'])
+        self.assertEqual('../miniseed/0407LHN.ms',
+                         ret[1]['raw_file_name_s'])
         ph5_object.ph5close()
-        os.remove('master.ph5')
-        os.remove('miniPH5_00001.ph5')
-        os.remove('metadatatoph5.log')
-        os.remove('datatoph5.log')
+        del_files_in_dir(self.tmpdir)
```
[the remark this answers is not in the document]
Comment (damhuonglan) [reply ph5/utilities/tests/test_obspytoph5.py]: It needs to clear files for a new test inside function's test (says inside test_main there are sub tests, each sub test need an empty tempdir. The deletion in tempdir class code only work when function's test finish.

[the remark this answers is not in the document]
Comment (dsentinel) [reply ph5/utilities/tests/test_obspytoph5.py]: Right, so this lends itself to each subtest being its own test.
This way all the file cleanup is done by the the setUp and tearDown and all the bookkeeping of what files are created can be removed.

[the remark this answers is not in the document]
Comment (damhuonglan) [reply ph5/utilities/tests/test_obspytoph5.py]: What do you mean by "lends itself". Do you mean create one test for each subtest?

[the remark this answers is not in the document]
Comment (dsentinel) [reply ph5/utilities/tests/test_obspytoph5.py]: Yes.

**Comment by dsentinel** [comment]
The docker tests are getting tempdir's mixed up:
`test_data/tmpEHBgy4` above doesn't match `test_data/tmptIEbcA` below

```
test_main (ph5.utilities.tests.test_metadatatoph5.TestMetadatatoPH5) ... ERROR
test_main has FAILED. Inspect files created in /home/travis/build/PIC-IRIS/PH5/ph5/test_data/tmpEHBgy4.
```
test_data/tmpEHBgy4 above doesn't match test_data/tmptIEbcA below
```
ERROR: test_main (ph5.utilities.tests.test_metadatatoph5.TestMetadatatoPH5)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/PIC-IRIS/PH5/ph5/utilities/tests/test_metadatatoph5.py", line 55, in test_main
    metadatatoph5.main()
  File "/home/travis/build/PIC-IRIS/PH5/ph5/utilities/metadatatoph5.py", line 492, in main
    LOGGER.warning("{0} not found. Creating...".format(ph5file))
  File "/home/travis/miniconda/envs/ph5/lib/python2.7/logging/__init__.py", line 1186, in warning
    self._log(WARNING, msg, args, **kwargs)
  File "/home/travis/miniconda/envs/ph5/lib/python2.7/logging/__init__.py", line 1293, in _log
    self.handle(record)
  File "/home/travis/miniconda/envs/ph5/lib/python2.7/logging/__init__.py", line 1303, in handle
    self.callHandlers(record)
  File "/home/travis/miniconda/envs/ph5/lib/python2.7/logging/__init__.py", line 1343, in callHandlers
    hdlr.handle(record)
  File "/home/travis/miniconda/envs/ph5/lib/python2.7/logging/__init__.py", line 766, in handle
    self.emit(record)
  File "/home/travis/miniconda/envs/ph5/lib/python2.7/logging/__init__.py", line 963, in emit
    self.stream = self._open()
  File "/home/travis/miniconda/envs/ph5/lib/python2.7/logging/__init__.py", line 950, in _open
    stream = open(self.baseFilename, self.mode)
IOError: [Errno 2] No such file or directory: '/home/travis/build/PIC-IRIS/PH5/ph5/test_data/tmptIEbcA/metadatatoph5.log'
```

The fix in the other branch only works because the same dir is being reused in all the tests

**Comment by dsentinel** [comment]
This is a bit of a mess:
`initialize_ex` is being called twice
and in metadatatoph5.py line 481 is `#debugging` comment and a `chdir` call

**Comment by damhuonglan** [comment]
Thank you abut the comment that reminds me to look at which tempdir is being used for which test function. I know what is happening now. Just don't know why it didn't happen on my local machine.

**Comment by dsentinel** [comment]
@damhuonglan I think the broken tests in docker are related to `FileHandlers`being attached to the `logger` being left open and attempted to be written to in subsequent tests, but by then the tmpdir is deleted.

`LogTestCase` should use `setUp` instead of `setUpClass` and destroy and rebuild the logger or remove all handlers for each test.

**Comment by damhuonglan** [comment]
@dsentinel LogTestCase is only related StreamHandler. Changing LogTestCase will not affect the behavior of FileHandler.

In metadatatoph5.py, same with other file, FileHandler with a specific path to log file is added (metadatatoph5.py line 484 to 489). You can see that each tool have their own log file. I need to remove it  at the end of main() so that another FileHandler with a different path to log file can be used (line 520).

If you want to use one FileHandler, which means one log file, for every tool it can be done in ph5/__init__. However users may have difficulties reading the log file. We should discuss with the data group be for doing that.

**Comment by dsentinel** [comment]
Please follow my last comment. No need to change `ph5/__init__.py` or any other sources outside of tests

**Comment by damhuonglan** [comment]
I have tried your suggestion on PR #387. Please have a look.

**Comment by damhuonglan** [comment]
For some reason, PR #387 doesn't check on docker. I have to push to this branch to test but it still not works. So I am testing on another PR #389. will ask for your review later.

**Comment by dsentinel** [comment]
Hold up for a bit with logging and tmpdir. I think I have a solution. I'll push a branch to test in docker and you can see the solution. On Mar 24, 2020 09:34, damhuonglan <notifications@github.com> wrote:
For some reason, PR #387 doesn't check on docker. I have to push to this branch to test but it still not works. So I am testing on another PR #389. will ask for your review later.

**Comment by dsentinel** [comment]
@damhuonglan Look at #388 for the solutions

- You found a way to remove the `fileHandler`'s, this has to be done in `tearDown` so any of the strange code/tests that create extra handlers are cleaned up.

- `PH5Availability` should use `TempDirTestCase` and `LogTestCase` it can inherit from both. See how I use super to follow the MRO. PH5Avaiability needs this as it creates files. It needs to use `os.path.join(self.home, 'ph5...')` to find the input files.

- A test pass, fail, or error, must return to homedir. Teardown isn't run if setup doesn't succeed. Use `addCleanup()` to ensure to return cwd to home.

Sorry this PR was really messy, and hard to pin down.

we also need to inherit from `TempDirTestCase` for the tests that are creating this files:
```
	extent.txt
	ph5/test_data/metadata/offset_t.kef
	ph5/test_data/metadata/sort_t.kef
	ph5/test_data/ph5/
```

Take what you want from #388 or merge it and clean it up. There's some debugging junk in there now.

**Comment by damhuonglan** [comment]
Changes
  + remove all handlers of files under ph5, except for the one in ph5/__itnit__.py that I have changed to StringIO for capturing log's messages
  + use addCleanup() to ensure to return cwd to home.
  + use super to follow the MRO

About inheriting from TempDirTestCase for the tests that are creating files and folder that you mentioned above, they are not created in unittests but were created for tests:
 + extent.txt and test are created by test_ph5availability.py. They are removed by letting the test inherit TempDirTestCase
 + ph5 is a folder created using the command bash created_ph5.sh in travis.yml. The ph5 files in this folder will be used for unittests run after that. If you need to remove that folder, we can remove in travis.yml. Let me know about it
+ ph5/test_data/metadata/offset_t.kef and ph5/test_data/metadata/sort_t.kef are created in created_ph5.sh when adding metadata to ph5.

**Comment by damhuonglan** [comment]
@dsentinel I updated the PR according to what I understand from your instruction.

At first all checks passed. But I push one more commit that delete a line to remove a file after a test because it will later be cleanup with TempDirTestCase.tearDown(). Then it keeps pending on Travis CI build. I think I need to commit something to ignite it checking again to remove the frozen point.

Do you think it is the correct solution?

**Comment by dsentinel** [comment]
@damhuonglan , tests have updated

**Comment by damhuonglan** [comment]
Thank you. Can you review the branch and check my comment right after the last commit please?

**Comment by damhuonglan** [comment]
> Take what you want from #388 or merge it and clean it up.
Sorry for copy from your branch. Because you said that I can take some from it show I just take the part that I need so I don't have to solve the conflict. If you want I can merge it again.

File: ph5/core/tests/test_base.py [hunk 13]
```diff
@@ -12,14 +16,35 @@ def initialize_ex(nickname, path, editmode=False):
...
+
 class TempDirTestCase(unittest.TestCase):
 
     def setUp(self):
         """
         create tmpdir
         """
         self.home = os.getcwd()
-        self.tmpdir = tempfile.mkdtemp() + "/"
+        self.tmpdir = tempfile.mkdtemp(dir=self.home + "/ph5/test_data/")
```
[the remark this answers is not in the document]
Comment (damhuonglan) [reply ph5/core/tests/test_base.py]: I will join that.

File: ph5/utilities/tests/test_metadatatoph5.py [hunk 14]
```diff
@@ -51,7 +50,7 @@ def test_main(self):
         test main function
         """
         testargs = ['metadatatoph5', '-n', 'master.ph5', '-f',
-                    'ph5/test_data/metadata/station.xml']
+                    '../metadata/station.xml']
```
[the remark this answers is not in the document]
Comment (damhuonglan) [reply ph5/utilities/tests/test_metadatatoph5.py]: This works because it passes the check. And actually when users run ph5 commands, they always use relative path. You can look in create_ph5.sh as an example.

I don't use os.join() because, as I told you, it creates the path that's greater than 32 character length in das_t's 'raw_file_name_s' and will be chop off at the end.

File: ph5/utilities/tests/test_metadatatoph5.py [hunk 15]
```diff
@@ -1,47 +1,46 @@
...
 
-        self.ph5_object = initialize_ex('master.ph5', self.path, True)
-        default_receiver_t = initialize_ph5.create_default_receiver_t()
-        initialize_ph5.set_receiver_t(default_receiver_t)
-        os.remove(default_receiver_t)
-        self.metadata = metadatatoph5.MetadatatoPH5(
-            self.ph5_object)
+    def tearDown(self):
+        if self._testMethodName != 'test_main':
+            self.ph5_object.ph5close()
```
[the remark this answers is not in the document]
Comment (damhuonglan) [reply ph5/utilities/tests/test_metadatatoph5.py]: The condision excludes test_main(). 

Anyway I think should put 'self.' in front of 'ph5_object' in test_main() so that I can clean it up in tearDown() and throw that condition away.

File: ph5/utilities/tests/test_metadatatoph5.py [hunk 16]
```diff
@@ -51,7 +50,7 @@ def test_main(self):
         test main function
         """
         testargs = ['metadatatoph5', '-n', 'master.ph5', '-f',
-                    'ph5/test_data/metadata/station.xml']
+                    '../metadata/station.xml']
         with patch.object(sys, 'argv', testargs):
             metadatatoph5.main()
         self.assertTrue(os.path.isfile('master.ph5'))
```
[the remark this answers is not in the document]
Comment (damhuonglan) [reply ph5/utilities/tests/test_metadatatoph5.py]: As the command in setUp(), it exclude test_main().

+ The ph5_object in setUp() is to provide the ph5_object for testing fuctions in metadatoph5.
+ The ph5_object in test_main() is different. Because main() includes create ph5_object and close it in itself. The ph5_object need to be created after main() added info and close it, so that we can use that ph5_object to read info for checking.

File: ph5/utilities/tests/test_metadatatoph5.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (dsentinel) [reply ph5/utilities/tests/test_metadatatoph5.py]: ok.. As this is an exception for some code that should be more robust. Put a comment stating so

**Comment by damhuonglan** [comment]
@dsentinel I want to merge your debug_tmpdir to this branch to give you credit for your code. However on top of that branch you said "Never merge, only to test docker." I am confused. Please let me know if I should merge debug_tmpdir to capture or not. Thank you.

**Comment by dsentinel** [comment]
When I wrote that I was hacking away, and put a lot of sloppy debugging. It's much cleaner now.

Thank you, but don't worry about merging for my credit at this point. Just keep in mind for the future.

**Comment by damhuonglan** [comment]
@dsentinel I think I addressed most of your requested changes except for the ph5_object initialized again in test_metadatatoph5.py and test_obspytoph5, and for using the relative path instead of os.path.join at some places in test_obspytoph5. I left comments in codes to explain why I have to keep it that way. Please review it.

In case you might miss some of my comments, offset_t.kef and sort_t.kef are not created by unittest. They are metadata for script created_ph5.sh to add metadata to ph5.

Folder ph5/ is created by created_ph5.sh in travis.yml to be used in unittest. I can remove it in travis.yml if you want to.

File: ph5/core/tests/test_base.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (dsentinel) [reply ph5/core/tests/test_base.py]: ...

File: ph5/core/tests/test_base.py [hunk 17]
```diff
@@ -12,14 +16,35 @@ def initialize_ex(nickname, path, editmode=False):
...
         os.chdir(self.tmpdir)
```
Comment (dsentinel) [inline ph5/core/tests/test_base.py:59]: Need to addcleanup() call... this is not ready, please give a self review

File: ph5/core/tests/test_base.py [hunk 18]
```diff
@@ -12,17 +16,52 @@ def initialize_ex(nickname, path, editmode=False):
...
         create tmpdir
         """
+        self.tmpdir = None
+        self.home = None
         self.home = os.getcwd()
-        self.tmpdir = tempfile.mkdtemp() + "/"
+        self.tmpdir = tempfile.mkdtemp(
+            dir=os.path.join(self.home, "ph5/test_data/"))
         os.chdir(self.tmpdir)
+        self.addCleanup(os.chdir, self.home)
```
Comment (damhuonglan) [inline ph5/core/tests/test_base.py:60]: addCleanup() is right here.

File: ph5/utilities/tests/test_metadatatoph5.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (damhuonglan) [reply ph5/utilities/tests/test_metadatatoph5.py]: Sorry. You are correct at this point. I misread this.

**Comment by damhuonglan** [comment]
@dsentinel : I have reviewed it myself according to your change request a couple of times. There are some changes I have made:
+ About the relative path in test_obspytoph5.py, I copy miniseed files to tmpdir and because they are in the same directory with master.ph5, obspytoph5 can read directly without path. So I don't have to use relative path or long path. Hope it solve the problem
+ I found a bug in test_base.TempDirTestCase that makes the code not delete tmpdir and report failed for all tests that follow a failed test in a unittest class. The reason is error list and failure list are for all test in a unittest class, not for each. To solve that I need a variable to keep track of the changes in the total number of errors and failures. I tried to start the variable in setUpClass() but failed because it reset that variable back to the starting value in each test. My solution is to start the variable total_error_failure at the top of the class. Hopefully you will agree with it.

File: ph5/utilities/tests/test_metadatatoph5.py [hunk 19]
```diff
@@ -1,47 +1,48 @@
...
     def setUp(self):
-        self.path = 'ph5/test_data/miniseedph5'
-        try:
-            os.mkdir(self.path)
-        except OSError:
-            print ("Creation of the directory %s failed" % self.path)
-        else:
-            print ("Successfully created the directory %s " % self.path)
+        super(TestMetadatatoPH5, self).setUp()
+        if self._testMethodName != 'test_main':
```
Comment (dsentinel) [inline ph5/utilities/tests/test_metadatatoph5.py:21]: Put `test_main` in another class without this setUp it doesn't need.

File: ph5/core/tests/test_base.py [hunk 20]
```diff
@@ -12,18 +16,57 @@ def initialize_ex(nickname, path, editmode=False):
...
+
 class TempDirTestCase(unittest.TestCase):
+    total_error_failure = 0
 
     def setUp(self):
         """
         create tmpdir
         """
+        self.tmpdir = None
+        self.home = None
```
Comment (dsentinel) [inline ph5/core/tests/test_base.py:56]: These are not needed.
This was for printing these variables before the first run.

File: ph5/clients/tests/test_ph5availability.py [hunk 21]
```diff
@@ -73,36 +63,25 @@ def checkFieldsMatch(fieldNames, fieldsList, dictList):
     return True
 
 
-class TestPH5Availability(unittest.TestCase):
-
+class TestPH5Availability(LogTestCase, TempDirTestCase):
     def setUp(self):
         """
         setup for tests
```
Comment (dsentinel) [inline ph5/clients/tests/test_ph5availability.py:69]: These kind of comments add no information. The function is already called `setup`.
Comments should add information / prevent code scanning to understand what the function does at a high level.
e.g.
creates and iniailtizes ph5 and availability object
creates a handle self.ph5test_path to the location of the ph5 test fixtures.
called before each test is run

File: ph5/utilities/tests/test_obspytoph5.py [hunk 22]
```diff
@@ -1,47 +1,61 @@
...
     def setUp(self):
-        self.path = 'ph5/test_data/miniseedph5'
+        super(TestObspytoPH5, self).setUp()
+        self.station_xml_path = os.path.join(
+            self.home, 'ph5/test_data/metadata/station.xml')
+        # copy miniseed to tmpdir
+        miniseed_dir = os.path.join(self.home, 'ph5/test_data/miniseed')
+        for filename in os.listdir(miniseed_dir):
+            file_path = os.path.join(miniseed_dir, filename)
+            shutil.copy(file_path, '.')
```
Comment (dsentinel) [inline ph5/utilities/tests/test_obspytoph5.py:27]: Do not copy the test files that are only read for every test.
In general, we don't want to copy. If a test writes to a file that should be created/copied to the tmpdir and then removed.

File: ph5/utilities/tests/test_obspytoph5.py [hunk 23]
```diff
@@ -1,47 +1,61 @@
...
+
+        if 'test_main' not in self._testMethodName:
```
Comment (dsentinel) [inline ph5/utilities/tests/test_obspytoph5.py:29]: These test_mains should be moved to another class

File: ph5/utilities/tests/test_metadatatoph5.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (damhuonglan) [reply ph5/utilities/tests/test_metadatatoph5.py:59]: I will create a different class for the test that does not need ph5_object.

File: ph5/core/tests/test_base.py [hunk 24]
```diff
@@ -12,18 +16,57 @@ def initialize_ex(nickname, path, editmode=False):
...
+        # put back the stream handler in ph5.__init__.logger
+        logger.handlers = cls.handlers
+
+    def tearDown(self):
+        # clean up handlers in any ph5 files' loggers that have been used
+        # but exclude logger from ph5/__init__.py to catch log's messages in
+        # next test
+        for k, v in logging.Logger.manager.loggerDict.items():
+            if (not isinstance(v, logging.PlaceHolder)) and ('ph5.' in k):
+                v.handlers = []
```
[the remark this answers is not in the document]
Comment (damhuonglan) [reply ph5/core/tests/test_base.py:44]: I just want to know your opinion about this. I will revert this.

File: ph5/utilities/tests/test_metadatatoph5.py (hunk 19 above)
Comment (damhuonglan) [reply ph5/utilities/tests/test_metadatatoph5.py:21]: will have another class for test_main.

File: ph5/clients/tests/test_ph5availability.py (hunk 21 above)
Comment (damhuonglan) [reply ph5/clients/tests/test_ph5availability.py:69]: I agree. You can see this kind of comment everywhere in unittests. I will remove the ones in tests that I have changed.

File: ph5/utilities/tests/test_obspytoph5.py (hunk 22 above)
Comment (damhuonglan) [reply ph5/utilities/tests/test_obspytoph5.py:27]: If you don't want to copy, I will revert to use relative path for miniseed files/folder.

File: ph5/utilities/tests/test_obspytoph5.py (hunk 23 above)
Comment (damhuonglan) [reply ph5/utilities/tests/test_obspytoph5.py:29]: will move test_mains to another class
